kotlin/nucleotide-count/src/main/kotlin/Dna.kt

```diff
@@ -0,0 +1,20 @@
+import java.lang.IllegalArgumentException
+
+class Dna(sequence: String) {
+
+    private val dnaSequence: String = sequence
+    private val validNucleotides = listOf('A', 'C', 'G', 'T')
+
+    init {
+        if (!dnaSequence.all { it in validNucleotides } && dnaSequence.isNotEmpty()) {
+            throw IllegalArgumentException()
+        }
+    }
+
+    val nucleotideCounts: Map<Char, Int>
+        get() {
```

Comment:
Usually in Kotlin, we try to use non-modifiable structures. I found a different way of doing the same:

```
private val baseNucleotideCount = mapOf(
        'A' to 0,
        'C' to 0,
        'G' to 0,
        'T' to 0
    )

    val nucleotideCounts: Map<Char, Int>
        get() {
            return baseNucleotideCount + dnaSequence.groupingBy { it }.eachCount()
        }

```

You can use `+` operator to add another Map's content to an existing one, and it creates a new one.

This will use an extra Map instead of collecting them in 1 map, like in your solution.
However, this is a better fit for Kotlin.

Also, please not the `groupingBy` and `eachCount` when processing the input.